enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,415 @@
+only use features which are supported on OCP and only those features which
+target Ingress resources. These changes may also be contributed back upstream
+since it may benefit other users.
+
+1. The annotation `alb.ingress.kubernetes.io/target-type` can be set to the
+value **_ip_** or **_instance_**. When the value is set to _instance_ the
+created load balancer will be configured to route traffic through a NodePort
+service. The _ip_ value only works when the cluster uses the Amazon VPC CNI and
+hence will not work on OCP/OKD clusters. The controller has to be modified so
+that this annotation is ignored and the controller always uses routing through
+_NodePort_ services.
+2. The _lb-controller_ has support for _Service_ type resources as
+well. When the annotation `service.beta.kubernetes.io/aws-load-balancer-type:
+"external"` is specified on a _Service_ resource of type _LoadBalancer_ the
+in-tree controller ignores this resource and the lb-controller instead
+provisions a Network Load Balancer with the correct configuration. Since we want
+to restrict the features supported by the controller to only Ingresses the
+lb-controller will have to be modified so that it does not reconcile any Service
+resources.
+
+#### Parallel operation  of the OpenShift router and lb-controller
+
+The OpenShift router manages ingresses that don’t have any ingress class value.
+This is the expected behavior currently and the lb-controller should not attempt
+to reconcile Ingresses which don’t have an ingress class set. If the user wants
+to switch over to using the lb-controller as default they can create an
+_IngressClass_ resource with the annotation
+`ingressclass.kubernetes.io/is-default-class` set to _true_ and the
+`spec.controller` value should be _ingress.k8s.aws/alb._ The user can also
+create multiple `IngressClass` resources with different configurations and have
+the `spec.controller` set to _ingress.k8s.aws/alb._ All Ingresses which belong
+to these classes will be reconciled by the lb-controller.
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+Other than standard unit testing the operator will have end-to-end tests for some common usage scenarios:
+
+1. Test which ensures that sufficient replicas are running and up-to-date.
+2. Test for Ingress with addon annotations but with the addons disabled.
+3. Test for Ingress with non-matching Ingress class. Verify that no load balancer is created.
+4. Test for when subnets are tagged manually. Verify that the load balancer is created only in the manually tagged subnet.
+
+### Graduation Criteria
+TBD
+
+#### Dev Preview -> Tech Preview
+TBD
+
+#### Tech Preview -> GA
+TBD
+
+#### Removing a deprecated feature
+NA
+
+### Upgrade / Downgrade Strategy
+TBD
+
+### Version Skew Strategy
+TBD
+
+### Operational Aspects of API Extensions
+TBD
+
+#### Failure Modes
+TBD
+
+#### Support Procedures
+TBD
+
+## Implementation History
+TBD
+
+## Drawbacks
+
+TBD
```

Comment:
For the "TBD" sections, do you plan to D them in a follow-up after this PR merges, or do you plan to D them in this PR?

Reply:
I've added details in some sections and will add more as we implement the operator. Some of the sections like Version Skew don't apply to this operator, so I've added an NA there.